soccer/planning/CollectPathPlanner.cpp

```diff
@@ -0,0 +1,295 @@
+    ShapeSet& obstacles = planRequest.obstacles;
+    vector<DynamicObstacle>& dynamicObstacles = planRequest.dynamicObstacles;
+
+    // Previous RRT path from last iteration
+    std::unique_ptr<Path>& prevPath = planRequest.prevPath;
+
+    // The small beginning part of the previous path
+    unique_ptr<Path> partialPath = nullptr;
+
+    // The is from the original robot position to the ball
+    // We only care about the replan lead time from the current pos in the path
+    // to the intercept point
+    RJ::Seconds timeIntoPreviousPath;
+
+    // How much of the future we are devoting to the partial path
+    // 0ms unless we have a partial path, then it's a partialReplanLeadTime
+    RJ::Seconds partialPathTime = 0ms;
+
+    // How much of the previous path to steal
+    const RJ::Seconds partialReplanLeadTime = RRTPlanner::getPartialReplanLeadTime();
+
+    // Change start instant to be the partial path end instead of the robot current location
+    // if we actually have already calculated a path the frame before
+    if (prevPath) {
+        timeIntoPreviousPath = curTime - prevPath->startTime();
+
+        // Make sure we still have time in the path to replan and correct
+        // since it's likely that the old path is slightly off
+        //
+        // ---|----------------|-----------------|
+        // TimeNow     EndPartialPrevPath  FinalTargetPoint
+        //                     |-----------------|
+        //          Amount of the path we can change this iteration
+        if (timeIntoPreviousPath < prevPath->getDuration() - 2*partialReplanLeadTime &&
+            timeIntoPreviousPath > 0ms) {
+            partialPath =
+                prevPath->subPath(0ms, timeIntoPreviousPath + partialReplanLeadTime);
+            partialPathTime = partialPath->getDuration() - timeIntoPreviousPath;
+            startInstant = partialPath->end().motion;
+        }
+    }
+
+    // Initialize the filter to the ball velocity so there's less ramp up
+    if (!averageBallVelInitialized) {
+        averageBallVel = ball.vel;
+        averageBallVelInitialized = true;
+    } else {
+        averageBallVel = *_targetPointAveragingGain * averageBallVel + 
+                         (1 - *_targetPointAveragingGain) * ball.vel;
+    }
+
+    // Do the transitions
+    float dist = (startInstant.pos - ball.pos).mag() - Robot_MouthRadius - *_approachDistTarget;
+    float speedDiff = (startInstant.vel - averageBallVel).mag() - *_touchDeltaSpeed;
+
+    // If we are close enough to the target point near the ball
+    // and almost the same speed we want, start slowing down
+    // TODO: Check for ball sense?
+    if (dist < *_distCutoffToControl && speedDiff < *_velCutoffToControl && currentState == Approach) {
+        currentState = Control;
+    }
+
+    // Check if we need to go back into approach
+    if ((ball.pos - startInstant.pos).mag() > 2*Robot_Radius && currentState == Control) {
+        cout << "Reseting" << endl;
```

Comment:
*resetting.

Also, can we not have `cout`s in the middle of random code? (Or could it be a little bit more specific as to what's happening?)